File: compiler/extensions/cpp/runtime/test/zserio/JsonEncoderTest.cpp
#include <cmath>
#include <sstream>

#include "gtest/gtest.h"
#include "zserio/JsonEncoder.h"

namespace zserio
{

TEST(JsonEncoderTest, encodeNull)
{
    std::ostringstream stream;
    JsonEncoder::encodeNull(stream);
    ASSERT_EQ("null", stream.str());
}

TEST(JsonEncoderTest, encodeBool)
{
    std::ostringstream stream;
    JsonEncoder::encodeBool(stream, true);
    ASSERT_EQ("true", stream.str());

    stream.str("");
    JsonEncoder::encodeBool(stream, false);
    ASSERT_EQ("false", stream.str());
}

TEST(JsonEncoderTest, encodeIntegral)
{
    std::ostringstream stream;
    JsonEncoder::encodeIntegral(stream, static_cast<uint8_t>(UINT8_MAX));
    ASSERT_EQ(std::to_string(UINT8_MAX), stream.str());
    stream.str("");
    JsonEncoder::encodeIntegral(stream, static_cast<uint16_t>(UINT16_MAX));
    ASSERT_EQ(std::to_string(UINT16_MAX), stream.str());
    stream.str("");
    JsonEncoder::encodeIntegral(stream, static_cast<uint32_t>(UINT32_MAX));
    ASSERT_EQ(std::to_string(UINT32_MAX), stream.str());
    stream.str("");
    JsonEncoder::encodeIntegral(stream, static_cast<uint64_t>(UINT64_MAX));
    ASSERT_EQ(std::to_string(UINT64_MAX), stream.str());

    stream.str("");
    JsonEncoder::encodeIntegral(stream, static_cast<int8_t>(INT8_MIN));
    ASSERT_EQ(std::to_string(INT8_MIN), stream.str());
    stream.str("");
    JsonEncoder::encodeIntegral(stream, static_cast<int8_t>(INT8_MAX));
    ASSERT_EQ(std::to_string(INT8_MAX), stream.str());
    stream.str("");
    JsonEncoder::encodeIntegral(stream, static_cast<int16_t>(INT16_MIN));
    ASSERT_EQ(std::to_string(INT16_MIN), stream.str());
    stream.str("");
    JsonEncoder::encodeIntegral(stream, static_cast<int16_t>(INT16_MAX));
    ASSERT_EQ(std::to_string(INT16_MAX), stream.str());
    stream.str("");
    JsonEncoder::encodeIntegral(stream, static_cast<int32_t>(INT32_MIN));
    ASSERT_EQ(std::to_string(INT32_MIN), stream.str());
    stream.str("");
    JsonEncoder::encodeIntegral(stream, static_cast<int32_t>(INT32_MAX));
    ASSERT_EQ(std::to_string(INT32_MAX), stream.str());
    stream.str("");
    JsonEncoder::encodeIntegral(stream, static_cast<int64_t>(INT64_MIN));
    ASSERT_EQ(std::to_string(INT64_MIN), stream.str());
    stream.str("");
    JsonEncoder::encodeIntegral(stream, static_cast<int64_t>(INT64_MAX));
    ASSERT_EQ(std::to_string(INT64_MAX), stream.str());

    stream.str("");
    JsonEncoder::encodeIntegral(stream, INT64_MIN);
    ASSERT_EQ("-9223372036854775808", stream.str());
    stream.str("");
    JsonEncoder::encodeIntegral(stream, -1000);
    ASSERT_EQ("-1000", stream.str());
    stream.str("");
    JsonEncoder::encodeIntegral(stream, 0);
    ASSERT_EQ("0", stream.str());
    stream.str("");
    JsonEncoder::encodeIntegral(stream, 1000);
    ASSERT_EQ("1000", stream.str());
    stream.str("");
    JsonEncoder::encodeIntegral(stream, UINT64_MAX);
    ASSERT_EQ("18446744073709551615", stream.str());
}

TEST(JsonEncoderTest, encodeFloatingPoint)
{
    std::ostringstream stream;
    JsonEncoder::encodeFloatingPoint(stream, -1.0);
    ASSERT_EQ("-1.0", stream.str());
    stream.str("");
    JsonEncoder::encodeFloatingPoint(stream, 0.0);
    ASSERT_EQ("0.0", stream.str());
    stream.str("");
    JsonEncoder::encodeFloatingPoint(stream, 1.0);
    ASSERT_EQ("1.0", stream.str());

    stream.str("");
    JsonEncoder::encodeFloatingPoint(stream, 3.5);
    ASSERT_EQ("3.5", stream.str());
    stream.str("");
    JsonEncoder::encodeFloatingPoint(stream, 9.875);
    ASSERT_EQ("9.875", stream.str());
    stream.str("");
    JsonEncoder::encodeFloatingPoint(stream, 0.6171875);
    ASSERT_EQ("0.6171875", stream.str());

    stream.str("");
    JsonEncoder::encodeFloatingPoint(stream, 1e20);
    ASSERT_TRUE("1e+20" == stream.str() || "1e+020" == stream.str())
            << "Value '" << stream.str() << "' does not match to neither '1e+20' nor '1e+020'";

    stream.str("");
    JsonEncoder::encodeFloatingPoint(stream, -1e+16);
    ASSERT_TRUE("-1e+16" == stream.str() || "-1e+016" == stream.str())
            << "Value '" << stream.str() << "' does not match to neither '-1e+16' nor '-1e+016'";

    stream.str("");
    JsonEncoder::encodeFloatingPoint(stream, static_cast<double>(NAN));
    ASSERT_EQ("NaN", stream.str());
    stream.str("");
    JsonEncoder::encodeFloatingPoint(stream, static_cast<double>(INFINITY));
    ASSERT_EQ("Infinity", stream.str());
    stream.str("");
    JsonEncoder::encodeFloatingPoint(stream, -static_cast<double>(INFINITY));
    ASSERT_EQ("-Infinity", stream.str());
}

TEST(JsonEncoderTest, encodeString)
{
    std::ostringstream stream;
    JsonEncoder::encodeString(stream, "");
    ASSERT_EQ("\"\"", stream.str());
    stream.str("");
    JsonEncoder::encodeString(stream, "test");
    ASSERT_EQ("\"test\"", stream.str());
    stream.str("");
    JsonEncoder::encodeString(stream, "München");
    ASSERT_EQ("\"München\"", stream.str());
    stream.str("");
    JsonEncoder::encodeString(stream, "€");
    ASSERT_EQ("\"€\"", stream.str());

    // escapes
    stream.str("");
    JsonEncoder::encodeString(stream, "\\");
    ASSERT_EQ("\"\\\\\"", stream.str());
    stream.str("");
    JsonEncoder::encodeString(stream, "\"");
    ASSERT_EQ("\"\\\"\"", stream.str());
    stream.str("");
    JsonEncoder::encodeString(stream, "\b");
    ASSERT_EQ("\"\\b\"", stream.str());
    stream.str("");
    JsonEncoder::encodeString(stream, "\f");
    ASSERT_EQ("\"\\f\"", stream.str());
    stream.str("");
    JsonEncoder::encodeString(stream, "\n");
    ASSERT_EQ("\"\\n\"", stream.str());
    stream.str("");
    JsonEncoder::encodeString(stream, "\r");
    ASSERT_EQ("\"\\r\"", stream.str());
    stream.str("");
    JsonEncoder::encodeString(stream, "\t");
    ASSERT_EQ("\"\\t\"", stream.str());

    stream.str("");
    JsonEncoder::encodeString(stream, "\n\t%^@(*aAzZ01234569$%^!?<>[]](){}-=+~:;/|\\\"\'Hello World2");
    ASSERT_EQ("\"\\n\\t%^@(*aAzZ01234569$%^!?<>[]](){}-=+~:;/|\\\\\\\"'Hello World2\"", stream.str());

    // <= 0x1F -> unicode escape
    stream.str("");
    JsonEncoder::encodeString(stream, "\x1F");
    ASSERT_EQ("\"\\u001f\"", stream.str());
}

} // namespace zserio

File: compiler/extensions/java/runtime/test/zserio/runtime/json/JsonEncoderTest.java
package zserio.runtime.json;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;

import org.junit.jupiter.api.Test;

public class JsonEncoderTest
{
    @Test
    public void encodeNull()
    {
        final StringWriter stringWriter = new StringWriter();
        try (final PrintWriter printWriter = new PrintWriter(stringWriter))
        {
            JsonEncoder.encodeNull(printWriter);
            assertEquals("null", stringWriter.toString());
        }
    }

    @Test
    public void encodeBoolean()
    {
        final StringWriter stringWriter = new StringWriter();
        try (final PrintWriter printWriter = new PrintWriter(stringWriter))
        {
            JsonEncoder.encodeBool(printWriter, true);
            assertEquals("true", stringWriter.toString());

            stringWriter.getBuffer().setLength(0);

            JsonEncoder.encodeBool(printWriter, false);
            assertEquals("false", stringWriter.toString());
        }
    }

    @Test
    public void encodeIntegral()
    {
        StringWriter stringWriter = new StringWriter();
        try (PrintWriter printWriter = new PrintWriter(stringWriter))
        {
            JsonEncoder.encodeIntegral(printWriter, 0xff);
            assertEquals(Integer.toString(0xff), stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeIntegral(printWriter, 0xffff);
            assertEquals(Integer.toString(0xffff), stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeIntegral(printWriter, 0xffffffff);
            assertEquals(Long.toString(0xffffffff), stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            final BigInteger uint64Max = BigInteger.ONE.shiftLeft(64).subtract(BigInteger.ONE);
            JsonEncoder.encodeIntegral(printWriter, uint64Max);
            assertEquals(uint64Max.toString(), stringWriter.toString());

            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeIntegral(printWriter, Byte.MIN_VALUE);
            assertEquals(Byte.toString(Byte.MIN_VALUE), stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeIntegral(printWriter, Byte.MAX_VALUE);
            assertEquals(Byte.toString(Byte.MAX_VALUE), stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeIntegral(printWriter, Short.MIN_VALUE);
            assertEquals(Short.toString(Short.MIN_VALUE), stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeIntegral(printWriter, Short.MAX_VALUE);
            assertEquals(Short.toString(Short.MAX_VALUE), stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeIntegral(printWriter, Integer.MIN_VALUE);
            assertEquals(Integer.toString(Integer.MIN_VALUE), stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeIntegral(printWriter, Integer.MAX_VALUE);
            assertEquals(Integer.toString(Integer.MAX_VALUE), stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeIntegral(printWriter, Long.MIN_VALUE);
            assertEquals(Long.toString(Long.MIN_VALUE), stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeIntegral(printWriter, Long.MAX_VALUE);
            assertEquals(Long.toString(Long.MAX_VALUE), stringWriter.toString());

            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeIntegral(printWriter, Long.MIN_VALUE);
            assertEquals("-9223372036854775808", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeIntegral(printWriter, -1000);
            assertEquals("-1000", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeIntegral(printWriter, 0);
            assertEquals("0", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeIntegral(printWriter, 1000);
            assertEquals("1000", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeIntegral(printWriter, uint64Max);
            assertEquals("18446744073709551615", stringWriter.toString());
        }
    }

    @Test
    public void encodeFloatingPoint()
    {
        StringWriter stringWriter = new StringWriter();
        try (PrintWriter printWriter = new PrintWriter(stringWriter))
        {
            JsonEncoder.encodeFloatingPoint(printWriter, -1.0);
            assertEquals("-1.0", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeFloatingPoint(printWriter, 0.0);
            assertEquals("0.0", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeFloatingPoint(printWriter, 1.0);
            assertEquals("1.0", stringWriter.toString());

            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeFloatingPoint(printWriter, 3.5);
            assertEquals("3.5", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeFloatingPoint(printWriter, 9.875);
            assertEquals("9.875", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeFloatingPoint(printWriter, 0.6171875);
            assertEquals("0.6171875", stringWriter.toString());

            // TODO[Mi-L@]: Note that scientific format is different from C++ and Python.
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeFloatingPoint(printWriter, 1e20);
            assertEquals("1.0E20", stringWriter.toString());

            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeFloatingPoint(printWriter, Double.NaN);
            assertEquals("NaN", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeFloatingPoint(printWriter, Float.NaN);
            assertEquals("NaN", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeFloatingPoint(printWriter, Double.POSITIVE_INFINITY);
            assertEquals("Infinity", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeFloatingPoint(printWriter, Float.POSITIVE_INFINITY);
            assertEquals("Infinity", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeFloatingPoint(printWriter, Double.NEGATIVE_INFINITY);
            assertEquals("-Infinity", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeFloatingPoint(printWriter, Float.NEGATIVE_INFINITY);
            assertEquals("-Infinity", stringWriter.toString());
        }
    }

    @Test
    public void encodeString()
    {
        StringWriter stringWriter = new StringWriter();
        try (PrintWriter printWriter = new PrintWriter(stringWriter);)
        {
            JsonEncoder.encodeString(printWriter, "");
            assertEquals("\"\"", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeString(printWriter, "test");
            assertEquals("\"test\"", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeString(printWriter, "München");
            assertEquals("\"München\"", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeString(printWriter, "€");
            assertEquals("\"€\"", stringWriter.toString());

            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeString(printWriter, "\\");
            assertEquals("\"\\\\\"", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeString(printWriter, "\"");
            assertEquals("\"\\\"\"", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeString(printWriter, "\b");
            assertEquals("\"\\b\"", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeString(printWriter, "\f");
            assertEquals("\"\\f\"", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeString(printWriter, "\n");
            assertEquals("\"\\n\"", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeString(printWriter, "\r");
            assertEquals("\"\\r\"", stringWriter.toString());
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeString(printWriter, "\t");
            assertEquals("\"\\t\"", stringWriter.toString());

            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeString(
                    printWriter, "\n\t%^@(*aAzZ01234569$%^!?<>[]](){}-=+~:;/|\\\"\'Hello World2");
            assertEquals("\"\\n\\t%^@(*aAzZ01234569$%^!?<>[]](){}-=+~:;/|\\\\\\\"'Hello World2\"",
                    stringWriter.toString());

            // <= 0x1F -> unicode escape
            stringWriter.getBuffer().setLength(0);
            JsonEncoder.encodeString(printWriter, new String(new byte[] {0x1F}, StandardCharsets.UTF_8));
            assertEquals("\"\\u001f\"", stringWriter.toString());
        }
    }
}
